cranelift/codegen/meta/src/shared/instructions.rs

```diff
@@ -4223,6 +4223,28 @@ pub(crate) fn define(
         .constraints(vec![WiderOrEq(Float.clone(), FloatTo.clone())]),
     );
 
+    ig.push(
+        Inst::new(
+            "fvdemote",
+            r#"
+                Convert `x` to a smaller floating point format.
+
+                Each lane in `x` is converted to the destination floating point format
+                by rounding to nearest, ties to even.
+
+                Cranelift currently only supports two floating point formats
+                - `f32` and `f64`. This may change in the future.
+
+                Fvdemote differs from fdemote in that with fvdemote it targets vectors.
+                With Fvdemote there is no contraint on having the result type have the same
+                number of vector lanes as the input type.
+                "#,
+            &formats.unary,
+        )
+        .operands_in(vec![x])
+        .operands_out(vec![a]),
```

Comment:
We need to make sure that we constrain this so that the following is not possible:
```
v1 = fconst.f64x2 ...
v2 = fvdemote.f32 v1
```
Right now this looks possible, but we shouldn't be able to demote from a vector to a scalar (doesn't make sense). If we add the `WiderOrEq(x, a)` constraint, though, the types allowed should be, e.g., `f64x2` -> `f32x2`, but `f32x2` isn't really a valid type in WebAssembly (it is in CLIF).

I think we should use the `.merge_lanes()` approach here so that the types are  `f64x2` -> `f32x4` and then in the instruction description **we have to specify that the top two lanes are zeroed out**. Or alternately, just explicitly put the types in.

Reply:
I've created new types with explicit lanes. Interesting enough after doing this I had to update the state push call in code_translator.rs. I also could have used split_lanes here though that would have been less constrained and semantically speaking just sounds vague on what that implies.